### PR TITLE
refactor(payment_methods): update `connector_mandate_details` for card metadata changes

### PR DESCRIPTION
**Pull request by Aprabhat19**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description
Allow update of `connector_mandate_details` for card in case of metadata changes(in case of expiry changes). First the  value is set as inactive and if a successful payment has been made then set it back to active

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?
 
- Nothreeds

> - Create a MA and a MCA
> - Make a off_Session save card payment
> - Make a off_Session payment again with the same card just and changes expiry 
> - After a successful payment the Connector Mandate Details would be overriden
Screenshot 2024-12-16 at 4 21 51 PM

- Follow the same step for 3threeds
> - Create a MA and a MCA
> - Make a off_Session save card payment
> - Make a off_Session payment again with the same card just and changed expiry 
> -Before a successful redirection ( the connector mandate is inactive)
Screenshot 2024-12-16 at 4 23 50 PM

> - After a successful redirection ( the connector mandate is active)
Screenshot 2024-12-16 at 4 24 24 PM

## Checklist

- [ ] I formatted the code `cargo +nightly fmt --all`
- [ ] I addressed lints thrown by `cargo clippy`
- [ ] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: crates/router/src/core/payments/tokenization.rs [hunk 1]
```diff
@@ -1252,3 +1280,42 @@ pub fn update_connector_mandate_details(
...
+pub fn update_connector_mandate_details_status(
+    merchant_connector_id: Option<id_type::MerchantConnectorAccountId>,
+    mandate_details: Option<diesel_models::PaymentsMandateReference>,
+) -> RouterResult<Option<serde_json::Value>> {
+    let mandate_reference = match mandate_details {
+        Some(mut payment_mandate_reference) => {
+            if let Some(mca_id) = merchant_connector_id {
+                payment_mandate_reference.entry(mca_id).and_modify(|pm| {
+                    let update_rec = diesel_models::PaymentsMandateReferenceRecord {
+                        connector_mandate_id: pm.connector_mandate_id.clone(),
```
Comment (Narayanbhat166) [inline crates/router/src/core/payments/tokenization.rs:1294]: you can just update the `connector_mandate_status` field only right?

File: crates/router/src/core/payments/tokenization.rs [hunk 2]
```diff
@@ -1252,3 +1280,42 @@ pub fn update_connector_mandate_details(
...
+                        payment_method_type: pm.payment_method_type,
+                        original_payment_authorized_amount: pm.original_payment_authorized_amount,
+                        original_payment_authorized_currency: pm
+                            .original_payment_authorized_currency,
+                        mandate_metadata: pm.mandate_metadata.clone(),
+                        connector_mandate_status: Some(ConnectorMandateStatus::Inactive),
```
[the remark this answers is not in the document]
Comment (Narayanbhat166) [reply crates/router/src/core/payments/tokenization.rs:1300]: yes, that way this function can be used to make some record active as well.

File: crates/router/src/core/payments/tokenization.rs [hunk 3]
```diff
@@ -1252,3 +1280,42 @@ pub fn update_connector_mandate_details(
...
     Ok(connector_mandate_details)
 }
+
+#[cfg(feature = "v1")]
+pub fn update_connector_mandate_details_status(
+    merchant_connector_id: Option<id_type::MerchantConnectorAccountId>,
+    mandate_details: Option<diesel_models::PaymentsMandateReference>,
+) -> RouterResult<Option<serde_json::Value>> {
+    let mandate_reference = match mandate_details {
+        Some(mut payment_mandate_reference) => {
```
Comment (Narayanbhat166) [inline crates/router/src/core/payments/tokenization.rs:1290]: use zip

File: crates/router/src/core/payments/tokenization.rs [hunk 4]
```diff
@@ -1252,3 +1280,42 @@ pub fn update_connector_mandate_details(
 
     Ok(connector_mandate_details)
 }
+
+#[cfg(feature = "v1")]
+pub fn update_connector_mandate_details_status(
+    merchant_connector_id: Option<id_type::MerchantConnectorAccountId>,
+    mandate_details: Option<diesel_models::PaymentsMandateReference>,
```
Comment (Narayanbhat166) [inline crates/router/src/core/payments/tokenization.rs:1287]: why this function takes both of these arguments as optional? the caller can decide this. Both arguments should be mandatory here

File: crates/router/src/core/payments/tokenization.rs [hunk 5]
```diff
@@ -458,7 +459,34 @@ where
                                 resp.payment_method_id = payment_method_id;
 
                                 let existing_pm = match payment_method {
-                                    Ok(pm) => Ok(pm),
+                                    Ok(pm) => {
+                                        let connector_mandate_details = update_connector_mandate_details_status(
+                                            merchant_connector_id,
+                                            pm
```
Comment (Narayanbhat166) [inline crates/router/src/core/payments/tokenization.rs:465]: Can you do this before calling the function?

File: crates/router/src/core/payments/tokenization.rs [hunk 6]
```diff
@@ -458,7 +459,34 @@ where
...
+                                            })
+                                            .transpose()
+                                            .change_context(errors::ApiErrorResponse::InternalServerError)
+                                            .attach_printable("Failed to deserialize to Payment Mandate Reference ")?
+                                        )?;
+                                        payment_methods::cards::update_payment_method_connector_mandate_details(
+                                            state,
+                                            key_store,
+                                            db,
+                                            pm.clone(),
```
Comment (Narayanbhat166) [inline crates/router/src/core/payments/tokenization.rs:481]: we can avoid cloning this whole object

File: crates/router/src/core/payments/tokenization.rs (hunk 1 above)
Comment (Aprabhat19) [reply crates/router/src/core/payments/tokenization.rs:1294]: Yea the others needs to be the same

File: crates/router/src/core/payments/tokenization.rs (hunk 6 above)
Comment (Aprabhat19) [reply crates/router/src/core/payments/tokenization.rs:481]: No clone is unavoidable here

**Review by Narayanbhat166 — APPROVED** [review]

**Review by Sarthak1799 — APPROVED** [review]

**Pull request merged**